### ISSUE TITLE
Show execution time in arangosh

### ISSUE THREAD

**Issue by matcho** [arangodb/arangodb#3516]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [x] 3.2

On this operating system:
- [x] Debian .deb

### this is a suggestion for arangosh
When running queries via arangosh, execution time is not shown. This makes it difficult to evaluate performance.

Aardvark does it, but it is not always available (network limitations), and one might prefer using arangosh in general.

Could it be possible to add an option to always show timings, as PostgreSQL's `psql` does with `\timing` for example ?

Or maybe there's already a way to do this ?

Thanks
Mathias

**Comment by hkernbach** [comment]
You can get the execution time via the arangosh. See following example:
`db._query('for u in users return u').getExtra();`

Result will be something like this:
```
{
  "stats" : {
    "writesExecuted" : 0,
    "writesIgnored" : 0,
    "scannedFull" : 4,
    "scannedIndex" : 0,
    "filtered" : 0,
    "httpRequests" : 0,
    "executionTime" : 0.0001888275146484375
  },
  "warnings" : [ ]
}
```

Where the `executionTime` property displays the execution time in seconds.

**Comment by jsteemann** [comment]
Suggested patch for automatic command execution time tracking:
```diff
diff --git a/arangosh/Shell/ConsoleFeature.cpp b/arangosh/Shell/ConsoleFeature.cpp
index 9343dab0f7..e2e5eb3d93 100644
--- a/arangosh/Shell/ConsoleFeature.cpp
+++ b/arangosh/Shell/ConsoleFeature.cpp
@@ -31,6 +31,7 @@
 #include "ProgramOptions/Section.h"
 #include "Shell/ClientFeature.h"
 
+#include <iomanip>
 #include <iostream>
 
 using namespace arangodb;
@@ -416,6 +417,14 @@ ConsoleFeature::Prompt ConsoleFeature::buildPrompt(ClientFeature* client) {
     if (esc) {
       if (c == '%') {
         result.push_back(c);
+      } else if (c == 't') {
+        std::ostringstream tmp;
+        tmp << std::setprecision(6) << std::fixed << TRI_microtime();
+        result.append(tmp.str());
+      } else if (c == 'p') {
+        std::ostringstream tmp;
+        tmp << std::setprecision(6) << std::fixed << _lastDuration;
+        result.append(tmp.str());
       } else if (c == 'd') {
         if (client != nullptr) {
           result.append(client->databaseName());
diff --git a/arangosh/Shell/ConsoleFeature.h b/arangosh/Shell/ConsoleFeature.h
index 7a5002c436..30de8a111c 100644
--- a/arangosh/Shell/ConsoleFeature.h
+++ b/arangosh/Shell/ConsoleFeature.h
@@ -78,6 +78,7 @@ class ConsoleFeature final : public application_features::ApplicationFeature {
   void closeLog();
   void log(std::string const&);
   void flushLog();
+  void setLastDuration(double duration) { _lastDuration = duration; }
 
   struct Prompt {
     std::string _plain;
@@ -106,6 +107,7 @@ class ConsoleFeature final : public application_features::ApplicationFeature {
   bool _supportsColors;
   FILE* _toPager;
   FILE* _toAuditFile;
+  double _lastDuration;
 };
 }
 
diff --git a/arangosh/Shell/V8ShellFeature.cpp b/arangosh/Shell/V8ShellFeature.cpp
index b948b27087..76ba45b435 100644
--- a/arangosh/Shell/V8ShellFeature.cpp
+++ b/arangosh/Shell/V8ShellFeature.cpp
@@ -349,8 +349,10 @@ int V8ShellFeature::runShell(std::vector<std::string> const& positionals) {
 
   bool const isBatch = isatty(STDIN_FILENO) == 0;
   bool lastEmpty = isBatch;
+  double lastDuration = 0.0;
 
   while (true) {
+    _console->setLastDuration(lastDuration);
     _console->setPromptError(promptError);
     auto prompt = _console->buildPrompt(client);
 
@@ -365,6 +367,7 @@ int V8ShellFeature::runShell(std::vector<std::string> const& positionals) {
     if (input.empty()) {
       promptError = false;
       lastEmpty = true;
+      lastDuration = 0.0;
       continue;
     }
     lastEmpty = isBatch;
@@ -392,10 +395,13 @@ int V8ShellFeature::runShell(std::vector<std::string> const& positionals) {
 
     // execute command and register its result in __LAST__
     v8LineEditor.setExecutingCommand(true);
+    double t1 = TRI_microtime();
 
     v8::Handle<v8::Value> v = TRI_ExecuteJavaScriptString(
         _isolate, context, TRI_V8_STD_STRING(_isolate, input), name, true);
 
+    lastDuration = TRI_microtime() - t1;
+
     v8LineEditor.setExecutingCommand(false);
 
     if (v.IsEmpty()) {
```
That would print command execution times as part of the ArangoShell prompt. Which is potentially quite nice. Will turn this into a proper feature should there be time.

**Comment by jsteemann** [comment]
There is PR 3555 (https://github.com/arangodb/arangodb/pull/3555) for this now.

**Comment by matcho** [comment]
Thank you !

**Comment by sleto-it** [comment]
Hi @matcho,

Many thanks for this feature request. A PR has been merged in the devel branch.  Option -console.prompt can be used. Eg.:

build/bin/arangosh --server.endpoint tcp://127.0.0.1:3333 -console.prompt "%p> "

0.000000> db._query('for d in test return d')
[object ArangoQueryCursor, count: 1, cached: false, hasMore: false]
[
  {
    "_key" : "126",
    "_id" : "test/126",
    "_rev" : "_V69ODWS--_",
    "a" : 1
  }
]
0.005693>

Can we close this ticket now, or is there anything else we can do to help?

Many thanks,

**Comment by matcho** [comment]
Hi @sleto-it 

Looks perfect, thanks a lot. Looking forward to see it released.
Unfortunately I don't really have time to test the develop branch at the moment, but I'm sure the feature is what I need.

All the best,
Mathias

**Comment by matcho** [comment]
Hi.
Cannot make it work with 3.2.9; the "%p" part of the prompt remains empty no matter which command is issued.

**Comment by jsteemann** [comment]
The patch has not be merged into the 3.2 branch, but into the 3.3 and devel branches.
That means the functionality will become available in an official release starting with version 3.3.
